[sw/silicon_creator] Add unit tests for sw/device/silicon_creator/otbn_util

jwnrt opened the pull request:
Adds unit tests for the four functions in `sw/device/silicon_creator/otbn_util`.

Since this file mostly just calls `sw/device/silicon_creator/lib/drivers/otbn.c` functions, these tests are heavily based on theirs.

engdoreis commented on sw/device/silicon_creator/lib/otbn_util_unittest.cc:
```diff
@@ -0,0 +1,176 @@
+  };
+
+  // Test assumption.
+  static_assert(OTBN_DMEM_SIZE_BYTES >= 4 * dmem_data.size(),
```
```suggestion
  static_assert(OTBN_DMEM_SIZE_BYTES >= sizeof(uint32_t) * dmem_data.size(),
```

alphan replied:
also applies to all other usages.

engdoreis commented on sw/device/silicon_creator/lib/otbn_util_unittest.cc:
```diff
@@ -0,0 +1,176 @@
+class OtbnWriteTest : public OtbnTest {};
+
+TEST_F(OtbnWriteTest, Success) {
+  const uint32_t dest = 6;
```
Prefer `constexpr` in C++, [see](https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#features).
Constants should be named like `kDest`, [see](https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#naming).
You can also use a unamed enum here, [see](https://docs.opentitan.org/doc/sg/c_cpp_coding_style/#constants-and-preprocessor-macros).

engdoreis reviewed (commented):
Thanks @jwnrt, I just left some comments.

alphan commented on sw/device/silicon_creator/lib/otbn_util_unittest.cc:
```diff
@@ -0,0 +1,176 @@
+  static_assert(OTBN_IMEM_SIZE_BYTES >= 4 * imem_data.size(),
+                "OTBN IMEM size too small");
+
+  otbn_t otbn;
+  otbn_init(&otbn);
```
nit: can you move these below just above the `EXPECT_EQ` at the end so that expectations and the code being tested are clearly separated? Applies to other test points as well.

jadephilipoom commented on sw/device/silicon_creator/lib/otbn_util_unittest.cc:
```diff
@@ -0,0 +1,176 @@
+                       });
+
+    EXPECT_ABS_READ32(base_ + OTBN_ERR_BITS_REG_OFFSET, err_bits);
+    EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
+
+    if (err_bits == kOtbnErrBitsNoError) {
+      EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
```
Yep, the extra read is for this check:
https://github.com/lowRISC/opentitan/blob/cbba276294efa163b5c5be029a9487c9908d9dd2/sw/device/silicon_creator/lib/drivers/otbn.c#L72-L73
Just a hardening measure, re-reading the register and checking it's definitely idle.

jadephilipoom reviewed (approved):
LGTM. It would also be nice to have a test that ensures loading or executing an app fails if OTBN is not idle (OTBN does a secure wipe on startup, so it's important that we check this), but that doesn't have to block this PR.

jwnrt commented:
Thanks all! Updated with style changes.

> It would also be nice to have a test that ensures loading or executing an app fails if OTBN is not idle

I've changed `OtbnLoadAppSuccess` to ensure it waits for OTBN to be idle before starting, but haven't added tests to check that each idle check throughout is handled correctly.

```diff
+ // `otbn_busy_wait_for_done` - begin with busy to ensure we wait until idle.
+ EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusBusyExecute);
+ EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusBusySecWipeDmem);
+ // Read twice for hardening.
+ EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
+ EXPECT_ABS_READ32(base_ + OTBN_STATUS_REG_OFFSET, kOtbnStatusIdle);
```

Also noticed I had tested `otbn_execute` by mistake instead of `otbn_execute_app`. Updated to use the right function and also added two cases:
- `OtbnExecuteBusy` to test OTBN not being idle after execution finishes is handled.
- `OtbnExecuteNotLoaded` to test `otbn_execute_app` being called before an `otbn_load_app`.

I can't find where `otbn_execute_app` checks for idle before sending the execute command, does it only check after?

jadephilipoom commented:
> I can't find where `otbn_execute_app` checks for idle before sending the execute command, does it only check after?

You're right -- looking at the code, I remember that we decided since ROM only loads one program, it's OK to just check for idle when loading the app and then check afterwards, so `otbn_execute_app` doesn't have a separate check.
